File: src/main/java/de/meshcloud/example/springkotlin/services/CostCalculationService.java
package de.meshcloud.example.springkotlin.services;

import de.meshcloud.example.springkotlin.exceptions.NotFoundException;
import de.meshcloud.example.springkotlin.model.Project;
import de.meshcloud.example.springkotlin.model.Resource;
import de.meshcloud.example.springkotlin.repositories.ProjectRepository;
import org.springframework.stereotype.Service;

import java.time.Period;
import java.time.temporal.ChronoUnit;

@Service
public class CostCalculationService {

    private ProjectRepository projectRepository;

    public CostCalculationService(ProjectRepository projectRepository) {
        this.projectRepository = projectRepository;
    }

    public Double calculateCosts(Long projectId) {
        Project project = projectRepository.findOne(projectId);
        if (project == null) {
            throw new NotFoundException("Project with id " + projectId + " does not exist");
        }

        return project.getResources().stream()
                .mapToDouble(r -> calculateResourceCost(r))
                .sum();
    }

    private double calculateResourceCost(Resource resource) {
        double cost = getPlatformCost(resource);
        cost = applyIllegalUseCost(resource, cost);
        return cost;
    }

    private double getPlatformCost(Resource resource) {
        double cost;
        /*
         * Simplified cost calculation. Should consider i.e. running VMs, networks, etc in OpenStack or
         * Apps and Services in Cloud Foundry, etc.
         */
        switch (resource.getType()) {
            case OPENSTACK:
                cost = 0.03;
                break;
            case CLOUDFOUNDRY:
                cost = 0.02;
                break;
            case KUBERNETES:
                cost = 0.01;
                break;
            default:
                throw new IllegalArgumentException("Resource Type '" + resource.getType() + "' not supported.");
        }
        return resource.durationInHours() * cost;
    }

    private double applyIllegalUseCost(Resource resource, double cost) {
        String description = resource.getProject().getDescription();
        if (description != null && description.contains("Bitcoin")) {
            return cost * 10;
        }
        return cost;
    }
}

File: src/main/kotlin/de/meshcloud/example/springkotlin/services/CostCalculationService.kt
package de.meshcloud.example.springkotlin.services

import de.meshcloud.example.springkotlin.model.Resource
import de.meshcloud.example.springkotlin.model.ResourceType
import de.meshcloud.example.springkotlin.repositories.ProjectRepository
import de.meshcloud.example.springkotlin.repositories.findOneOrThrow
import org.springframework.stereotype.Service

@Service
class CostCalculationService(private val projectRepository: ProjectRepository) {

  fun calculateCosts(projectId: Long): Double {
    val project = projectRepository.findOneOrThrow(projectId)

    return project.resources
        .map { calculateResourceCost(it) }
        .sum()
  }

  private fun calculateResourceCost(resource: Resource) =
      getPlatformCost(resource).let {
        applyIllegalUseCost(resource, it)
      }

  private fun getPlatformCost(resource: Resource): Double {
    /*
     * Simplified cost calculation. Should consider i.e. running VMs, networks, etc in OpenStack or
     * Apps and Services in Cloud Foundry, etc.
     */
    val cost: Double = when (resource.type) {
      ResourceType.OPENSTACK -> 0.03
      ResourceType.CLOUDFOUNDRY -> 0.02
      ResourceType.KUBERNETES -> 0.01
    }
    
    return resource.durationInHours() * cost
  }

  private fun applyIllegalUseCost(resource: Resource, cost: Double): Double {
    val description = resource.project.description
    return if (description?.contains("Bitcoin") == true) {
      cost * 10
    } else cost
  }
}
